snapcraft/internal/build_providers/images.py

```diff
@@ -0,0 +1,137 @@
+        self.snap_arch = snap_arch
+        self.url = url
+        self.checksum = checksum
+        self.algorithm = algorithm
+
+        self._image_cache = FileCache(taxonomy="build-images-{}".format(self.base))
+
+    def _download_and_cache(self) -> str:
+        request = requests.get(self.url, stream=True, allow_redirects=True)
+        if not request.ok:
+            raise errors.BuildImageRequestError(
+                base=self.base, status_code=request.status_code
+            )
+        # First create the prefix as tempfile.TemporaryDirectory does not do that for you
+        os.makedirs(self._image_cache.file_cache, exist_ok=True)
+        with tempfile.TemporaryDirectory(
+            prefix=self._image_cache.file_cache
+        ) as tmp_dir:
+            download_file = os.path.join(tmp_dir, "{}-vm".format(self.base))
+            download_requests_stream(request, download_file)
+            calculated_digest = calculate_hash(download_file, algorithm=self.algorithm)
+            if self.checksum != calculated_digest:
+                raise errors.BuildImageChecksumError(
+                    expected=self.checksum,
+                    calculated=calculated_digest,
+                    algorithm=self.algorithm,
+                )
+            return self._image_cache.cache(
+                filename=download_file, algorithm=self.algorithm, hash=self.checksum
+            )
+
+    def get(self):
+        cached_file = self._image_cache.get(
+            hash=self.checksum, algorithm=self.algorithm
+        )
+        if not cached_file:
+            cached_file = self._download_and_cache()
+        # TODO verify nothing is using this as a backing store before implementing.
+        # image_cache.prune(keep_hash=image.checksum)
+        return cached_file
+
+
+def _get_build_images(base: str) -> Dict[str, _Image]:
+    if base == "core16":
+        return dict(
+            amd64=_Image(
+                base="core16",
+                snap_arch="amd64",
+                url="https://cloud-images.ubuntu.com/releases/16.04/release-20180703/ubuntu-16.04-server-cloudimg-amd64-disk1.img",  # noqa: E501
+                checksum="79549e87ddfc61b1cc8626a67ccc025cd7111d1af93ec28ea46ba6de70819f8c",  # noqa: E501
+                algorithm="sha256",
+            )
+        )
+    elif base == "core18":
+        return dict(
+            amd64=_Image(
+                base="core18",
+                snap_arch="amd64",
+                url="https://cloud-images.ubuntu.com/releases/18.04/release-20180724/ubuntu-18.04-server-cloudimg-amd64.img",  # noqa: E501
+                checksum="6d663a8fd5eddd916f4aef4fd06d0f7f4cf0bb191f170b8c84cd2adf297bc5c3",  # noqa: E501
+                algorithm="sha256",
+            )
+        )
+    else:
+        raise KeyError(base)
+
+
+def setup(*, base: str, snap_arch: str, size: str, image_path: str) -> None:
+    """Setup an instance for base and snap_arch on sized size image_path."""
```

Comment:
I'm not parsing this description: "... on sized size image_path" ? Also, mind adding sphinx docs?

Reply:
s/on/of/ :man_shrugging: